File: src/main/python/sorting/mergesort/merge_sort.py
def merge_sort(to_sort):
    merge_sort_helper(to_sort, 0, len(to_sort))


def merge_sort_helper(to_sort, left, right):
    if left >= right - 1:
        return

    mid = int((left + right) / 2)
    merge_sort_helper(to_sort, left, mid)
    merge_sort_helper(to_sort, mid, right)
    merge(to_sort, to_sort[left: mid], to_sort[mid: right], left)


def merge(to_sort, left, right, start_pos):
    i = 0
    j = 0
    k = start_pos

    while i < len(left) or j < len(right):
        if i >= len(left):
            to_sort[k] = right[j]
            j += 1
            k += 1
        elif j >= len(right):
            to_sort[k] = left[i]
            i += 1
            k += 1
        else:
            if left[i] < right[j]:
                to_sort[k] = left[i]
                i += 1
                k += 1
            else:
                to_sort[k] = right[j]
                j += 1
                k += 1


if __name__ == "__main__":
    arr = [1, 2, 6, 4, 5]
    print(str(arr))
    merge_sort(arr)
    print(str(arr))

File: src/main/java/sorting/mergesort/MergeSort.java
package sorting.mergesort;

import java.util.Arrays;

public class MergeSort {

  public static void sort(int[] arr) {
    sort(arr, 0, arr.length);
  }

  private static void sort(int[] arr, int left, int right) {
    if (left >= right - 1) {
      return;
    }

    int mid = (left + right) / 2;
    sort(arr, left, mid);
    sort(arr, mid, right);

    int[] leftSide = new int[mid - left];
    int[] rightSide = new int[right - mid];

    System.arraycopy(arr, left, leftSide, 0, leftSide.length);
    System.arraycopy(arr, mid, rightSide, 0, rightSide.length);

    merge(arr, leftSide, rightSide, left);
  }

  private static void merge(int[] arr, int[] left, int[] right, int startPos) {
    int k = startPos;

    for (int i = 0, j = 0; i < left.length || j < right.length; ) {
      if (i >= left.length) {
        arr[k++] = right[j++];
      } else if (j >= right.length) {
        arr[k++] = left[i++];
      } else {
        if (left[i] < right[j]) {
          arr[k++] = left[i++];
        } else {
          arr[k++] = right[j++];
        }
      }
    }
  }

  public static void main(String[] args) {
    int[] arr = {1, 5, 3, 5, 3, 2, 4, 5, 2};
    System.out.println(Arrays.toString(arr));
    sort(arr);
    System.out.println(Arrays.toString(arr));
  }
}
